crates/iroha_core/src/peers_gossiper.rs

```diff
@@ -0,0 +1,151 @@
+            .into_iter()
+            .map(|peer| (peer.id, peer.address))
+            .collect();
+        let gossiper = Self {
+            initial_peers,
+            gossip_peers: HashMap::new(),
+            network,
+        };
+        gossiper.network_update_peers_addresses();
+
+        let (message_sender, message_receiver) = mpsc::channel(1);
+        (
+            PeersGossiperHandle { message_sender },
+            Child::new(
+                tokio::task::spawn(gossiper.run(message_receiver, shutdown_signal)),
+                OnShutdown::Abort,
+            ),
+        )
+    }
+
+    async fn run(
+        mut self,
+        mut message_receiver: mpsc::Receiver<PeersGossip>,
+        shutdown_signal: ShutdownSignal,
+    ) {
+        let mut gossip_period = tokio::time::interval(Duration::from_secs(60));
+        loop {
+            tokio::select! {
+                _ = gossip_period.tick() => {
+                    self.gossip_peers()
+                }
+                _ = self.network.wait_online_peers_update(|_| ()) => {
+                    self.gossip_peers();
+                }
+                Some(peers_gossip) = message_receiver.recv() => {
+                    self.handle_peers_gossip(peers_gossip);
+                }
+                () = shutdown_signal.receive() => {
+                    iroha_logger::debug!("Shutting down peers gossiper");
+                    break;
+                },
+            }
+            tokio::task::yield_now().await;
+        }
+    }
+
+    fn gossip_peers(&self) {
+        let online_peers = self.network.online_peers(Clone::clone);
+        let online_peers = UniqueVec::from_iter(online_peers);
+        let data = NetworkMessage::PeersGossiper(Box::new(PeersGossip(online_peers)));
+        self.network.broadcast(Broadcast { data });
+    }
+
+    fn handle_peers_gossip(&mut self, PeersGossip(peers): PeersGossip) {
+        for peer in peers {
+            self.gossip_peers.insert(peer.id, peer.address);
+        }
+        self.network_update_peers_addresses();
+    }
+
+    fn network_update_peers_addresses(&self) {
+        let online_peers = self.network.online_peers(Clone::clone);
+        let online_peers_ids = online_peers
+            .into_iter()
+            .map(|peer| peer.id)
+            .collect::<HashSet<_>>();
+
+        let mut peers = Vec::new();
+        for (id, address) in &self.initial_peers {
+            if !online_peers_ids.contains(id) {
+                peers.push((id.clone(), address.clone()));
+            }
+        }
+        for (id, address) in &self.gossip_peers {
+            if !online_peers_ids.contains(id) {
+                peers.push((id.clone(), address.clone()));
+            }
+        }
+
+        let update = UpdatePeers(peers);
+        self.network.update_peers_addresses(update);
+    }
+}
+
+/// Message for gossiping peers addresses.
+#[derive(Decode, Encode, Debug, Clone)]
+pub struct PeersGossip(UniqueVec<Peer>);
```

Comment:
```suggestion
pub struct PeersGossip(Vec<Peer>);
```

I'd rather verify that there are no duplicates in the message received from the other peer and refuse to decode the message then deduplicate